scm/app.go

```diff
@@ -0,0 +1,124 @@
+package scm
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+
+	"github.com/beatlabs/github-auth/app"
+	"github.com/beatlabs/github-auth/key"
+	"github.com/google/go-github/v43/github"
+)
+
+const (
+	AppEnv           = "APP_ID"
+	KeyEnv           = "APP_KEY"
+	SecretEnv        = "APP_SECRET"
+	KeyPath          = "APP_KEYPATH"
+	InstallationsAPI = "https://api.github.com/app/installations"
+	GitHubUserAPI    = "https://api.github.com/user"
+)
+
+// GithubAppConfig keeps parameters of the GitHub app
+type GithubAppConfig struct {
+	appID    string
+	clientID string
+	secret   string
+	keyPath  string
+}
+
+type GithubApp struct {
+	App    *app.Config
+	scms   *Scms
+	config *GithubAppConfig
+}
+
+func newAppConfig() *GithubAppConfig {
+	return &GithubAppConfig{
+		appID:    os.Getenv(AppEnv),
+		clientID: os.Getenv(KeyEnv),
+		secret:   os.Getenv(SecretEnv),
+		keyPath:  os.Getenv(KeyPath),
+	}
+}
+
+// Valid ensures that all configuration fields are not empty
+func (conf *GithubAppConfig) Valid() bool {
+	return conf.appID != "" && conf.keyPath != "" &&
+		conf.clientID != "" && conf.secret != ""
+}
+
+// AppClient creates client for the Quickfeed GitHub Application
+// This client can only access the metadata of the Application itself
+// To access organizations via GitHub API we need to derive an installation client
+// from this Application client for each course organization
+func NewApp() (*GithubApp, error) {
+	config := newAppConfig()
+	if !config.Valid() {
+		return nil, fmt.Errorf("error configuring GitHub App: %+v", config)
+	}
+	appKey, err := key.FromFile(config.keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("error reading key from file: %s", err)
+	}
+	appClientConfig, err := app.NewConfig(config.appID, appKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating GitHub application client: %s", err)
+	}
+	return &GithubApp{
+		config: config,
+		App:    appClientConfig,
+		scms:   NewScms(),
+	}, nil
+}
+
+// Creates a new scm client with access to the course organization
+func (ghApp *GithubApp) NewInstallationClient(ctx context.Context, courseOrg string) (*github.Client, error) {
+	resp, err := ghApp.App.Client().Get(InstallationsAPI)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching installations for GitHub app %s: %s", ghApp.config.appID, err)
```

Comment:
Here we should use `%w` instead of `%s` to wrap the error. (same for Lines 87, 91 and 105)...